Initial Adoption of Continous Deployment using semantic-release

keplersj opened the pull request:
To ease the modification/contribution to publish pipeline there has been some discussion of adopting a continuous deployment system for AtomLinter packages. To accomplish this we are making use of [`semantic-release`](https://github.com/semantic-release/semantic-release) and [`@semantic-release/apm-config`](https://github.com/semantic-release/apm-config).

This pull request has been [dynamically created using a script](https://gist.github.com/keplersj/8bc7622ea741c0964d12842bfc679c5d). While the result is not perfect, it does accompish most of the grunt work of adopting continous deployment. There is some reconciliation that needs to be done before this can be merged.

Among the things needed to be reconciled, an [APM Token](https://atom.io/account) and [GitHub Token](https://help.github.com/articles/creating-a-personal-access-token-for-the-command-line/) will need to be added to this repo's continous integration (most likely [Travis CI](https://docs.travis-ci.com/user/environment-variables#Defining-Variables-in-Repository-Settings)) configuration for automated deployments to work.

Again, this Pull Request has been created by a script made by @keplersj. Please mention him if something has gone wrong, and he'll be happy to help.

ref: AtomLinter/Meta#18

cc: @Arcanemagus

keplersj committed:
build(release): initial adoption of semantic-release and commitlint

ericcornelissen commented:
I added the tokens (APM and GitHub) to the Travis CI environment and it seems to me no further changes are required for this particular package.

I'm however feeling some pretty essential information is missing for people (like me) who are not as familiar with continuous deployment, @keplersj. I already [asked](https://github.com/AtomLinter/Meta/issues/18#issuecomment-381397654) about the names of the tokens which are not given (although they are pretty straightforward, them not being specified can cause a lot of doubt). But I'm also not entirely sure about how this system will work. So here is my take on understanding it:

> As I understand it the additions to the CI configuration automatically create a new release (under the `deploy` key in the Travis CI configuration) whenever a commit is made (I'm guessing only to `master` since `branches: only: - master`?). And the release notes are then generated based on the commit messages on `master`. But these commit messages must have a specific format in order for it to be able to do that, which [husky](https://www.npmjs.com/package/husky) will ensure.

Is that description more or less correct? Either way I would recommend adding such a description to the main issue in the AtomLinter/Meta repository so it is easy to find for everyone.

But that's just a recommendation, other then that good job on doing all this work @keplersj. I'm looking forward to using Continuous Deployment for the first time :grinning:

keplersj commented:
You've essentially got it! [Husky](https://github.com/typicode/husky) is a git-hook manager for npm projects. On the commit message hook we use [commitlint](https://github.com/marionebl/commitlint) to ensure the commit message follows the [conventional commits standard](https://conventionalcommits.org/). On every push to master (not necessarily dictated by the `branches` block, but by the `deploy` block) we run semantic-release which analyzes all of the commits since the last tag, determines if a release is ready, and if it is creates a changelog and publishes to apm.

ericcornelissen commented:
Clear, thank you for the clarification!

keplersj commented:
@ericcornelissen I'd like to apologize. I gave you the wrong environment variable name for the APM token in the other threat. It should be `ATOM_ACCESS_TOKEN` not `APM_TOKEN`

ericcornelissen commented:
Thanks for the update 👍

Arcanemagus commented:
@ericcornelissen Thanks for the suggestion, I've turned https://github.com/AtomLinter/Meta/issues/18#issue-254770705 into a massive FAQ/HOWTO for this 😉.

ericcornelissen committed:
Update Travis configuration

Based on https://github.com/AtomLinter/linter-js-yaml/blob/295b4865fb7af44f6b3e8446bba6c26cd35a17a9/.travis.yml

Arcanemagus commented:
Hmmm, Travis CI seems to have failed to build the merge of your commit, and instead is reporting the status for the initial commit for yours.

The current "state of the art" can be found in [AtomLinter/linter-stylelint:.travis.yml@ddf630770](https://github.com/AtomLinter/linter-stylelint/blob/ddf630770b0ed97be595a66bd3fa549844ab4876/.travis.yml), do you want me to update it to that (or you can)? A new commit _should_ get Travis CI to figure out it's mistake.

ericcornelissen committed:
Update Travis configuration 

Based on https://github.com/AtomLinter/linter-stylelint/blob/ddf630770b0ed97be595a66bd3fa549844ab4876/.travis.yml

ericcornelissen commented:
I updated it again @Arcanemagus, the previous update was inline with the [linter-js-yaml `.travis.yml` file](https://github.com/AtomLinter/linter-js-yaml/blob/295b4865fb7af44f6b3e8446bba6c26cd35a17a9/.travis.yml) you referenced in the FAQ/HOWTO... It seems to be working now, only the last two commits aren't formatted correctly 😅

Arcanemagus commented:
Make sure you run `npm install` so the new `git` hooks are installed that tell you about those problem before they hit CI 😉.

ericcornelissen commented:
I understand, I just made these changes through the GitHub web editor without really think about it :disappointed: It doesn't hurt to still merge this PR right?

Arcanemagus commented:
> It doesn't hurt to still merge this PR right?

Other than merging failed CI builds.... no. I'll squash and merge it so these commits don't break the `master` build (hopefully).

Arcanemagus commented:
:tada: This PR is included in version 1.1.0 :tada:

The release is available on [GitHub release](https://github.com/AtomLinter/linter-htmllint/releases/tag/v1.1.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
